File: algorithm/001-two_sum/ruby/solution.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
  complements = {}
  nums.each_with_index {
    |n, i|
    c = target - n
    complements[c] ? (return [complements[c], i]) : complements[n] = i
  }
end

File: algorithm/001-two_sum/java/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
  public int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> complements = new HashMap<>();
    for (int i = 0; i < nums.length; i++) {
      int c = target - nums[i];
      if (complements.containsKey(c)) {
        return new int[]{complements.get(c), i};
      } else {
        complements.put(nums[i], i);
      }
    }
    return new int[0];
  }
}